### PR TITLE
Common: Blocking Loop (extracted from Fifo.cpp)

### PR DESCRIPTION
**Pull request by degasus**
This PR tries to extract the synchronized main loop of our fifo as shared file within Common.

The new Common Class represent a way to implement a "busy loop" which sleeps on no work. It is optimized for a high frequent pushing of new data while the loop is already running.

### CODE REVIEW

**Comment by degasus** [comment]
degasus> so from a testing point of view, PR 2470 is fine?
JMC47> yes
JMC47> no change in functionality at all

File: Source/Core/Common/BlockingLoop.h [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+			return;
+
+		m_shutdown.Set();
+		Set();
+
+		if (block)
+			Wait();
+	}
+
+	bool IsRunning()
```
Comment (lioncash) [inline Source/Core/Common/BlockingLoop.h]: This can be a const function.

File: Source/Core/Common/BlockingLoop.h [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+ * It's a thread-safe way to trigger a new iteration without busy loops.
+ * It's optimized for high-usage iterations which usually are already runing while it's triggered often.
+ */
+class BlockingLoop
+{
+public:
+	BlockingLoop()
+	{
+		m_running.Clear();
+	};
```
Comment (lioncash) [inline Source/Core/Common/BlockingLoop.h]: semicolon doesn't need to be here.

File: Source/Core/Common/BlockingLoop.h [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+#include "Common/Event.h"
+#include "Common/Flag.h"
+
+namespace Common
+{
+
+/*
+ * This class provides a synchronized loop.
+ * It's a thread-safe way to trigger a new iteration without busy loops.
+ * It's optimized for high-usage iterations which usually are already runing while it's triggered often.
```
Comment (Tilka) [inline Source/Core/Common/BlockingLoop.h]: *running

(And use `//` comments only.)

File: Source/Core/Common/BlockingLoop.h [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	~BlockingLoop()
+	{
+		Stop();
+	}
+
+	/*
+	 * Triggers to rerun the payload of the Run() function at least once again.
+	 * This function will never block and is designed to run as fast as possible.
```
Comment (Tilka) [inline Source/Core/Common/BlockingLoop.h]: You probably mean s/run/finish/ ?

File: Source/Core/Common/BlockingLoop.h [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+		m_is_pending.Set();
+		m_shutdown.Clear();
+		m_may_sleep.Set();
+	}
+
+	/*
+	 * Mainloop of this object.
+	 * The payload callback is called at least as often as it's needed to match the Set() requirements.
+	 */
+	template <class F> void Run(F&& payload)
```
Comment (Tilka) [inline Source/Core/Common/BlockingLoop.h]: Nitpick: `template<class F>`

File: Source/Core/VideoCommon/Fifo.cpp [hunk 6]
```diff
@@ -370,30 +362,14 @@ void RunGpuLoop()
 				AsyncRequests::GetInstance()->PullEvents();
 			}
 
+			// We are likely shuting down, so use our free time to do as much work as possible.
```
Comment (Tilka) [inline Source/Core/VideoCommon/Fifo.cpp]: *shutting

File: Source/Core/Common/BlockingLoop.h [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+		if (!m_running.TestAndSet())
+			return;
+		m_is_pending.Set();
+		m_shutdown.Clear();
+		m_may_sleep.Set();
+	}
+
+	// Mainloop of this object.
+	// The payload callback is called at least as often as it's needed to match the Set() requirements.
+	template<class F> void Run(F&& payload)
```
Comment (randomstuff) [inline Source/Core/Common/BlockingLoop.h]: `Run(F payload)`?

Comment (degasus) [inline Source/Core/Common/BlockingLoop.h]: So you want to run the copy operator implicitly? As far as I know, this is the same as:
´´´c++
F tmp = your_payload;
Run(tmp); // with my declaration
´´´

Comment (randomstuff) [inline Source/Core/Common/BlockingLoop.h]: This is for consistency with the library standard C++ API with always passes functors by value, such as `lock(condition_variable<mutex> lock, Predicate pred)`.

Comment (degasus) [inline Source/Core/Common/BlockingLoop.h]: I wanted the reference to be able to use any object later (aka state of this "lambda"), but whatever...

File: Source/UnitTests/Common/BlockingLoopTest.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+		});
+
+		// Now Wait must block.
+		loop.Prepare();
+
+		// The payload must run at least once on startup.
+		loop.Wait();
+		EXPECT_EQ(signaled.load(), received.load());
+
+		for (int j = 0; j < 100; j++)
```
Comment (degasus) [inline Source/UnitTests/Common/BlockingLoopTest.cpp]: Run this loop threaded to see inter-thread race conditions.

File: Source/Core/Common/BlockingLoop.h [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+		m_is_running.Clear();
+		m_is_pending.Clear();
+		m_running.Clear();
+
+		m_done_event.Set();
+	}
+
+	// Quits the mainloop.
+	// By default, it will wait until the Mainloop did quit.
```
Comment (phire) [inline Source/Core/Common/BlockingLoop.h]: ...until the Mainloop quits.

File: Source/Core/Common/BlockingLoop.h [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+		}
+	}
+
+	// Half start the worker.
+	// So this object is in running state and Wait() will block until the worker calls Run().
+	// This may be called from any thread and is supposed to call at least once before Wait() is used.
+	void Prepare()
+	{
+		// There is a race condition if the other threads is still in calling this function
+		// while the loop thread is initializing. Using this lock will ensure a valid state.
```
Comment (phire) [inline Source/Core/Common/BlockingLoop.h]: I think this should be, "... if the other threads call this function while the..." not 100% sure.

File: Source/Core/VideoCommon/Fifo.cpp [hunk 11]
```diff
@@ -370,30 +362,14 @@ void RunGpuLoop()
 				AsyncRequests::GetInstance()->PullEvents();
 			}
 
+			// We are likely shutting down, so use our free time to do as much work as possible.
```
Comment (phire) [inline Source/Core/VideoCommon/Fifo.cpp]: How about " // The fifo is empty and it's unlikely we will get any more work in the near future. Make sure VertexManager finishes drawing any primitives it has stored in it's buffer."

**Comment by comex** [comment]
Okay.  I think I've convinced myself that the atomic logic is correct, so sorry for doubting you @degasus :)

```
Proof that Wait() does not return early:
    - Flag's operations are done with std::memory_order_seq_cst, so the following linear idea of time is correct:
    - once Set ensures that m_is_running is true, there must be moments when (a) running is clear and (b) pending is clear, in that order, for a subsequent Wait to return
        - m_is_running is only cleared by Run while m_is_pending is true, so m_is_pending must subsequently be cleared after that clear (which itself is after Set started, so the work is already reflected in the other atomics)
        - m_is_pending will not be cleared until after the /next/ payload invocation
    - (alternately, m_running could become false, but this is documented to cause Wait to return)

Proof that Wait() does return:
    - The only ways for the clear-running -> payload -> clear-pending -> set-done-event path in Run to not eventually make both remain false are if:
        - m_is_running is constantly set by another Set on a different thread - that could cause problems in a multiple-producer scenario, but with Dolphin, Wait/Set are only called from the CPU thread - in any case, it will at least return once activity subsides
        - m_shutdown is set.  In this case, m_running is cleared before m_done_event is set, so Wait will not block.
    - Since m_done_event is always set at the end of this path, Wait cannot miss the timing.

Suggestions:
    - There is a 100ms timeout in Run's wait that isn't obviously necessary.  I guess this is needed by the GPU thread for some event loop nonsense, but it should probably be customizable in this class.
    - The names of the variables are confusing and not really accurate - e.g. m_is_running can be false while payload is running.  (I partially blame this for my confusion, but of course atomics are hard and I'm dumb :)
    - Do we care about the multiple-producer issue?
```

**Comment by degasus** [comment]
@comex Thanks for your huge efforts, I've tried to implement your suggestions.
My two atomic<bool> variables are now within one atomic<int> which is controlled by an enum. I hope the new name is better.
The timeout is now optional.
I've added some more comments to describe what I'm doing.

Missing:
- I didn't care about the multiple-producer issue. I think it's possible to solve this, but I'm not sure if it's worth.
- I'm not sure if it's worth to add a new "state" for sleeping. So we could skip the event notification while the worker thread is in the busy loop. Do you think this is worth?

**Comment by degasus** [comment]
JMC47> seems exactly the same

**Comment by comex** [comment]
Haven't thought about the new version as thoroughly as the last one, but it seems reasonable.  Thanks!

**Pull request merged**